### PR TITLE
Enumerable test

### PR DESCRIPTION
**Pull request by gonjavi**

### CODE REVIEW

File: spec/enumerable_spec.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+RSpec.describe Enumerable do 
+  let(:array) { [4,5,6] }
+  let(:my_each_arr) { array.my_each { |n| n } } 
+  myindex = Proc.new { [4,5,6].my_each_with_index { |val,index| index } } 
+  mapy = Proc.new { [4,5,6].my_map {|num| num * 10 } }
+  selecty = Proc.new { [4,5,6].my_select { |n| n.even? } }
+
+  describe "#my_each" do 
+    it "it tests my each" do 
+      expect(my_each_arr).to eq(array)
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: This test needs to account for positive and negative outcomes, in this case you are checking that two of your let blocks are equal, that is a problem because with the block that you are passing you are not testing the behavior of your method, because the method being called is just returning the indexes that are being iterated. You should be testing the behaviour directly for example, checking if your method does not modify the array, and checking if your method is able to interact with the elements being iterated over and not just returning the number, also you should be using your functions in your let block, ideally you should only need the array variable.

File: spec/enumerable_spec.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+		end
+  end
+  
+  describe "#my_each_with_index" do 
+    it "it tests my each with index" do
+      expect(array.my_each_with_index { |x, i| array[i] = i }).to eql([0, 1, 2])
```
Comment (jofer86) [inline spec/enumerable_spec.rb:19]: Please add a test that accounts for negative scenarios, like the example I gave you on the previous comment.

File: spec/enumerable_spec.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+
+  describe "#my_count" do
+    it "it tests my count" do
+      expect(array.my_count { |x| x } ).to eq(3)
+    end
+  end
+
+  describe "#my_map" do
+    it "it tests my map" do 
+      expect(mapy.call).to eq([40, 50, 60])
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: You should be doing the testing here instead of using the call method, this is what testing is for, your method should be here and not declared in a variable at top.

File: spec/enumerable_spec.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+RSpec.describe Enumerable do 
+  let(:array) { [4,5,6] }
+  let(:my_each_arr) { array.my_each { |n| n } } 
+  myindex = Proc.new { [4,5,6].my_each_with_index { |val,index| index } } 
+  mapy = Proc.new { [4,5,6].my_map {|num| num * 10 } }
+  selecty = Proc.new { [4,5,6].my_select { |n| n.even? } }
+
+  describe "#my_each" do 
+    it "it tests my each" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please be more descriptive here.

File: spec/enumerable_spec.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  selecty = Proc.new { [4,5,6].my_select { |n| n.even? } }
+
+  describe "#my_each" do 
+    it "it tests my each" do 
+      expect(my_each_arr).to eq(array)
+		end
+  end
+  
+  describe "#my_each_with_index" do 
+    it "it tests my each with index" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please be more descriptive here.

File: spec/enumerable_spec.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      expect(array.my_each_with_index { |x, i| array[i] = i }).to eql([0, 1, 2])
+    end
+  end
+
+  describe "#my select" do
+    it "it test my select" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please be more descriptive here.

File: spec/enumerable_spec.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  describe "#my select" do
+    it "it test my select" do 
+      expect([4,5,6].my_select { |n| n.even? } ).to eq([4,6] )
+    end
+  end
+
+  describe "#my_all?" do 
+    it "it tests my all?" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please be more descriptive here.

File: spec/enumerable_spec.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      expect([4,5,6].my_select { |n| n.even? } ).to eq([4,6] )
+    end
+  end
+
+  describe "#my_all?" do 
+    it "it tests my all?" do 
+      expect(array.my_all? { |x| x > 3 }).to eq(true)
+    end
+
+    it "it tests my all?" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please be more descriptive here.

File: spec/enumerable_spec.rb [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+  describe "#my_all?" do 
+    it "it tests my all?" do 
+      expect(array.my_all? { |x| x > 3 }).to eq(true)
+    end
+
+    it "it tests my all?" do 
+      expect(array.my_all? { |x| (Numeric) }).to eq(true)
+    end
+
+    it "it tests my all?" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please be more descriptive here.

File: spec/enumerable_spec.rb [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      expect(array.my_all? { |x| (Numeric) }).to eq(true)
+    end
+
+    it "it tests my all?" do 
+      expect(array.my_all? { |x| x > 7 } ).to eq(false)
+    end
+  end
+
+  describe "#my_any" do
+    it "it tests if any greater than 7" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please do describe a general behavior, this is a bit too explicit and does not describe in detail the behavior being tested.

File: spec/enumerable_spec.rb [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+      expect(array.my_any? { |x| x > 7 } ).to eq(false)
+    end
+
+    it "it tests if any equal to 4" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please do describe a general behavior, this is a bit too explicit and does not describe in detail the behavior being tested.

File: spec/enumerable_spec.rb [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+      expect(array.my_any? { |x| x == 4 } ).to eq(true)
+    end
+
+  end
+
+  describe "#my_none" do 
+    it "it tests if none greater than 8" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please do describe a general behavior, this is a bit too explicit and does not describe in detail the behavior being tested.

File: spec/enumerable_spec.rb [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      expect(array.my_none? { |x| x > 8 } ).to eq(true)
+    end
+
+    it "it tests if none less than 5" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please do describe a general behavior, this is a bit too explicit and does not describe in detail the behavior being tested.

File: spec/enumerable_spec.rb [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+      expect(array.my_none? { |x| x < 5 } ).to eq(false)
+    end
+  end
+
+  describe "#my_count" do
+    it "it tests my count" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please be more descriptive here.

File: spec/enumerable_spec.rb [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+      expect(array.my_count { |x| x } ).to eq(3)
+    end
+  end
+
+  describe "#my_map" do
+    it "it tests my map" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please be more descriptive here.

File: spec/enumerable_spec.rb [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  describe "#my_map" do
+    it "it tests my map" do 
+      expect(mapy.call).to eq([40, 50, 60])
+    end
+  end
+
+  describe "#my_inject" do 
+    it "it tests my inject" do
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: Please be more descriptive here.

File: spec/enumerable_spec.rb [hunk 17]
```diff
@@ -0,0 +1,80 @@
+require "./enumerable.rb"
+
+RSpec.describe Enumerable do 
+  let(:array) { [4,5,6] }
+  let(:my_each_arr) { array.my_each { |n| n } }
```
Comment (jofer86) [inline spec/enumerable_spec.rb]: From this line up to line 8 are not necesary and should be removed, because we need the tests to be done in the test blocks and not pass the values we got from a variable declaration.

**Review by jofer86 — CHANGES_REQUESTED** [review]
# Great Effort!!!! :clap: :clap: Although we need to make a few adjustments to the code.

Firstly I need to request that your `it` functions are a bit more descriptive, because we have to assume that our code is gonna be read by other people and it makes your test suite more understandable when you describe at some detail the behaviour that you are testing, please do the required changes where it is being asked.

In general your testing suite works, we just need to account for positive and negative scenerios where applicable, please do follow up this instruction for all your tests cases, also in your .rspec file if you add this line `--format documentation` you get a better output in your terminal, this change is not required but it makes the output of your tests more readable.

Once you are done please ask for a new code review.

File: spec/enumerable_spec.rb (hunk 1 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb]: perfect

File: spec/enumerable_spec.rb (hunk 2 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb:19]: perfect

File: spec/enumerable_spec.rb (hunk 3 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb]: perfect

File: spec/enumerable_spec.rb (hunk 4 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb]: perfect

File: spec/enumerable_spec.rb (hunk 5 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb]: perfect

File: spec/enumerable_spec.rb (hunk 13 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb]: perfect

File: spec/enumerable_spec.rb (hunk 14 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb]: perfect

File: spec/enumerable_spec.rb (hunk 15 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb]: perfect

File: spec/enumerable_spec.rb (hunk 16 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb]: perfect

File: spec/enumerable_spec.rb (hunk 17 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb]: perfect

File: spec/enumerable_spec.rb [hunk 18]
```diff
@@ -2,78 +2,87 @@
...
 
-    it "it tests if none less than 5" do
+    it "it tests if none value of the array is less than 5, if it is not correct return false, in this case there is a 4, it should return false" do
       expect(array.my_none? { |x| x < 5 } ).to eq(false)
     end
   end
 
   describe "#my_count" do
-    it "it tests my count" do
+    it "it counts all the numbers of the array, there are 3 numbers, it should return 3" do
```
Comment (Oghenebrume50) [inline spec/enumerable_spec.rb:77]: There are no negative test cases here

File: spec/enumerable_spec.rb [hunk 19]
```diff
@@ -2,78 +2,87 @@
...
       expect(array.my_count { |x| x } ).to eq(3)
     end
   end
 
   describe "#my_map" do
```
Comment (Oghenebrume50) [inline spec/enumerable_spec.rb:86]: No negative test case

File: spec/enumerable_spec.rb [hunk 20]
```diff
@@ -2,78 +2,87 @@
...
-    it "it tests my map" do 
-      expect(mapy.call).to eq([40, 50, 60])
+    it "it iterates each number on the array and multiplies it by 10 and returns the array with the new values" do 
+      expect([4,5,6].my_map {|num| num * 10 } ).to eq([40, 50, 60])
     end
   end
 
   describe "#my_inject" do
```
Comment (Oghenebrume50) [inline spec/enumerable_spec.rb:96]: No negative test here

**Review by Oghenebrume50 — CHANGES_REQUESTED** [review]
Please fix those few things and submit for re-review :+1:

File: spec/enumerable_spec.rb (hunk 18 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb:77]: ok perfect

File: spec/enumerable_spec.rb (hunk 19 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb:86]: perfect

File: spec/enumerable_spec.rb (hunk 20 above)
Comment (gonjavi) [reply spec/enumerable_spec.rb:96]: perfect

**Comment by maelfosso** [comment]
# Goob job :+1: 

Please go ahead and merge your branch.
Good luck with the next project

**Review by maelfosso — APPROVED** [review]
# Goob job :+1: 

Please go ahead and merge your branch.
Good luck with the next project
